server/src/main/java/org/elasticsearch/search/query/PartialHitCountCollector.java

```diff
@@ -43,26 +45,101 @@ public ScoreMode scoreMode() {
 
     @Override
     public LeafCollector getLeafCollector(LeafReaderContext context) throws IOException {
-        if (numCollected >= totalHitsThreshold) {
-            earlyTerminateIfNeeded();
-        }
+        earlyTerminateIfNeeded();
         return new FilterLeafCollector(super.getLeafCollector(context)) {
             @Override
             public void collect(int doc) throws IOException {
-                if (++numCollected > totalHitsThreshold) {
-                    earlyTerminateIfNeeded();
-                }
+                earlyTerminateIfNeeded();
+                hitsThresholdChecker.incrementHitCount();
                 super.collect(doc);
             }
         };
     }
 
     private void earlyTerminateIfNeeded() {
-        earlyTerminated = true;
-        throw new CollectionTerminatedException();
+        if (hitsThresholdChecker.isThresholdReached()) {
+            earlyTerminated = true;
+            throw new CollectionTerminatedException();
+        }
     }
 
     boolean hasEarlyTerminated() {
         return earlyTerminated;
     }
+
+    abstract static class HitsThresholdChecker {
```

Comment:
nit: could this and the two implementations be private? Doesn't seem to be used in tests, but maybe you have future usage in mind?

Reply:
good point. I think that I will extract this out to a top-level class as a follow-up because QueryPhaseCollectorManager has exactly the same.